panoptes_aggregation/panoptes/userify.py

```diff
@@ -0,0 +1,99 @@
+from collections import Iterable
+from json import dumps as jsonify
+from os import environ
+from flask import request
+from panoptes_client import Panoptes, User
+import requests
+from sys import modules
+
+class ConfigurationError(Exception):
+    pass
+
+this = modules[__name__]
+
+this.known_params = [
+    'destination'
+]
+
+this.connected = False
+
+this.users = {}
+
+this.destinations = {
+    'mast': 'https://mast-forward.zooniverse.org/',
+    'mockable': 'https://demo1580318.mockable.io/mast'
+}
+
+def userify():
+    if request.method == 'GET':
+        return 'Usage'
+    elif request.method in ['POST', 'PUT']:
+        find_fields = _discover_fields(request.args)
+        target_object = request.get_json()
+        destination = request.args.get('destination')
+
+        _stuff_object(target_object, find_fields)
+
+        if destination:
+            _forward_contents(target_object, destination)
+
+        return jsonify(target_object)
+
+def _forward_contents(payload, destination):
+    if destination not in this.destinations:
+        raise ConfigurationError('Unknown destination')
+
+    requests.post(url=this.destinations[destination], json=payload)
+    return
+
+def _stuff_object(target_object, find_fields):
+    for key in target_object.keys():
+        if type(target_object[key]) is dict:
+            _stuff_object(target_object[key], find_fields)
+
+    user_ids = _discover_user_ids(target_object)
+    for user_id in user_ids:
+        user = _retrieve_user(user_id)
+        target_object['users'] = target_object.get('users', [])
+        target_object['users'].append(_build_user_hash(user,find_fields))
+
+def _discover_fields(request_args):
+    raw_list = request_args.to_dict().keys()
+    return [ i for i in raw_list if i not in this.known_params ]
+
+def _discover_user_ids(target_object):
+    user_ids = []
+    if 'user_ids' in target_object:
+        user_ids = target_object['user_ids']
```

Comment:
Should this also be `.append()` like it is below?

Reply:
I think you can avoid the `_flatten` step by using list concatenation instead of appending:

```python
list_1 = [1, 2, 3]
list_2 = [4, 5, 6]
concat_list = list_1 + list_2
print(concat_list)  # will be [1, 2, 3, 4, 5, 6]
```

or if you have a  list-of-lists and want to concat them all you can call `sum`:

```python
list_1 = [1, 2, 3]
list_2 = [4, 5, 6]
list_of_list = [list_1, list_2]
concat_list = sum(list_of_list, [])  # the second argument is the "zero" value of the operation
print(concat_list)  # will be [1, 2, 3, 4, 5, 6]
```

So in this case:

```python
user_ids = []
if 'user_ids' in target_object:
    user_ids += sum(target_object['user_ids'], [])
if 'user_id' in target_object:
    user_ids += target_object['user_id']
```

Reply:
i don't think it needs to be "append" because before this it's an empty list. the reason i have to flatten isn't because of how i concatenate, but because i need to be able to handle things like `user_ids=[[3, 4], [2, 5]]` so flattening it is just the simplest way to make sure i have a flat list

Reply:
The API should define the type of variables you want and respond with invalid / raise an error if the supplied input doesn't match your expected format.

E.g only take `user_ids` and they must be in a flat array of min length 1, `user_ids: [1]`

I would drop the `user_id` support below and consolidate on this.